vendor-extra/LiberoPageBundle/src/Controller/PageController.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Libero\LiberoPageBundle\Controller;
+
+use Libero\LiberoPageBundle\Event\CreatePageEvent;
+use Libero\LiberoPageBundle\Event\LoadPageEvent;
+use Libero\LiberoPageBundle\Exception\NoContentSet;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Twig\Environment;
+use function count;
+use function Libero\LiberoPageBundle\text_direction;
+
+final class PageController
+{
+    private $dispatcher;
+    private $template;
+    private $twig;
+
+    public function __construct(Environment $twig, string $template, EventDispatcherInterface $dispatcher)
+    {
+        $this->twig = $twig;
+        $this->template = $template;
+        $this->dispatcher = $dispatcher;
+    }
+
+    public function __invoke(Request $request) : Response
+    {
+        $context = [
+            'lang' => $request->getLocale(),
+            'dir' => text_direction($request->getLocale()),
+        ];
+
+        $loadEvent = new LoadPageEvent($request, $context);
+        $this->dispatcher->dispatch($loadEvent::NAME, $loadEvent);
+
+        $createEvent = new CreatePageEvent($request, $loadEvent->getDocuments()->wait(), $loadEvent->getContext());
+        $this->dispatcher->dispatch($createEvent::NAME, $createEvent);
+
+        if (0 === count($createEvent->getContent())) {
+            throw NoContentSet::forPage($request->attributes->get('libero_page'));
+        }
+
+        return new Response(
+            $this->twig->render(
+                $this->template,
+                $createEvent->getContext() + [
```

Comment:
looks strange to merge the `context` (which is dynamic) with a fixed set of keys residing in the same object. Maybe Twig requires this structure, but given changes in the template it could be made nested such as:
```
[
    'context' => ...,
    'title' => ...,
    'context' => ...,
]
```

Reply:
This isn't new https://github.com/libero/browser/blob/5d8b471cab69b20b93997291ef41d46722eb22e1/vendor-extra/ContentPageBundle/src/Controller/ContentController.php#L65 but could (independently) be improved. Really just need the `lang` and `dir`, so probably better to just pick them out.